# CS 583-C-Homework 1
## Neural networks

---

### ***Fill your details below***
### Name: Corey Heckel
### CWID: 10462028
### Email ID: checkel
### References: ***Cite your references here***


---
### Submission guidelines:

#### 1. Submit this notebook along with its PDF version. You can do this by clicking File->Print->"Save as PDF"

#### 2. Name the file as "<mailID_HWnumber.extension>".  

For example, mailID is abcdefg @stevens.edu then name the files as abcdefg_HW1.ipynb and abcdefg_HW1.pdf.

#### 3. Please do not Zip your files.

---

In [1]:
#@title Installing Pytorch

!pip install torch
!pip install torchvision

  Using cached typing_extensions-4.12.2-py3-none-any.whl (37 kB)
  Attempting uninstall: typing-extensions
    Found existing installation: typing-extensions 4.1.1
    Uninstalling typing-extensions-4.1.1:
      Successfully uninstalled typing-extensions-4.1.1


In [2]:
#@title Import Dependencies

import torch
import torch.nn as nn
import torchvision.datasets as dsets
import torchvision.transforms as transforms
from torch.autograd import Variable

In [3]:
#@title Define Hyperparameters

input_size = 784 # img_size = (28,28) ---> 28*28=784 in total
hidden_size = 500 # number of nodes at hidden layer
num_classes = 10 # number of output classes discrete range [0,9]
num_epochs = 5 # number of times which the entire dataset is passed throughout the model
batch_size = 100 # the size of input data took for one iteration
lr = 1e-3 # size of step

In [4]:
#@title Downloading MNIST data

train_data = dsets.MNIST(root = './data', train = True,
                        transform = transforms.ToTensor(), download = True)

test_data = dsets.MNIST(root = './data', train = False,
                       transform = transforms.ToTensor())

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|███████████████████████████████████████████████████████████████████| 9912422/9912422 [00:06<00:00, 1563450.48it/s]


Extracting ./data\MNIST\raw\train-images-idx3-ubyte.gz to ./data\MNIST\raw

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|███████████████████████████████████████████████████████████████████████| 28881/28881 [00:00<00:00, 1314092.71it/s]


Extracting ./data\MNIST\raw\train-labels-idx1-ubyte.gz to ./data\MNIST\raw

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|███████████████████████████████████████████████████████████████████| 1648877/1648877 [00:00<00:00, 9686068.14it/s]


Extracting ./data\MNIST\raw\t10k-images-idx3-ubyte.gz to ./data\MNIST\raw

Failed to download (trying next):
HTTP Error 403: Forbidden



100%|██████████████████████████████████████████████████████████████████████████| 4542/4542 [00:00<00:00, 919693.38it/s]

Extracting ./data\MNIST\raw\t10k-labels-idx1-ubyte.gz to ./data\MNIST\raw



In [5]:
#@title Loading the data

train_gen = torch.utils.data.DataLoader(dataset = train_data,
                                             batch_size = batch_size,
                                             shuffle = True)

test_gen = torch.utils.data.DataLoader(dataset = test_data,
                                      batch_size = batch_size,
                                      shuffle = False)

## Q1. (50 points): design a neural network, provide justification.

PyTorch neural network documentation: https://pytorch.org/docs/stable/nn.html

In [6]:
#@title Define model class

class Net(nn.Module):
  def __init__(self, input_size, hidden_size, num_classes):
    super(Net,self).__init__()
    ''' code to build the model '''
    self.fc1 = nn.Linear(input_size, hidden_size)
    self.relu = nn.ReLU()
    self.fc2 = nn.Linear(hidden_size, num_classes)


  def forward(self,x):
    ''' code to define the forward pass '''
    output = self.fc1(x)
    output = self.relu(output)
    output = self.fc2(output)
    return output

self.fc1 represents the first fully connected hidden layer going from the input size (784 neurons) to the hidden size (500 neurons). ReLU is used as the activation function due to its ability to introduce non-linearity and efficiency. There is then a second fully connected layer which goes from the hidden size(500 neurons) to then the number of classes(10 neurons). The forward method passes the input data through the first fully connected layer, activated the ReLU function, passes the output through the second fully connected layer, resulting in a prediction of the class the input data. 

In [9]:
#@title Build the model

net = Net(input_size, hidden_size, num_classes)
if torch.cuda.is_available():
  net.cuda()

In [10]:
#@title Define loss-function & optimizer

loss_function = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam( net.parameters(), lr=lr)

##Q2.

##a) In the code provided below, What is the meaning of the "loss.backward()" step? Please explain the functionality. (15 pts)

This step computes the gradients of the loss function with respect to the weights and bias. This indicated how to adjust the parameters to reduce loss.  This is called backpropagation where the error is propagated backward through the network.

##b) What is the meaning of "optimizer.step()" and what does it do? (15 pts)

After computing the gradients in the previous step, this updates the model parameters using the calculated gradients to minimize the loss. Ultimately, this is used to optimize the parameters of the model.

In [11]:
#@title Training the model

for epoch in range(num_epochs):
  for i ,(images,labels) in enumerate(train_gen):
    images = Variable(images.view(-1,28*28))
    labels = Variable(labels)

    optimizer.zero_grad()
    outputs = net(images)
    loss = loss_function(outputs, labels)
    loss.backward() #Q2a (15 points): What is the meaning of this step? Please explain the functionality.
    optimizer.step() #Q2b (15 points): What is the meaning of this step? Please explain the functionality.

    if (i+1) % 100 == 0:
      print('Epoch [%d/%d], Step [%d/%d], Loss: %.4f'%(epoch+1, num_epochs, i+1, len(train_data)//batch_size, loss.item()))

Epoch [1/5], Step [100/600], Loss: 0.3856
Epoch [1/5], Step [200/600], Loss: 0.3547
Epoch [1/5], Step [300/600], Loss: 0.2332
Epoch [1/5], Step [400/600], Loss: 0.1790
Epoch [1/5], Step [500/600], Loss: 0.1084
Epoch [1/5], Step [600/600], Loss: 0.1370
Epoch [2/5], Step [100/600], Loss: 0.1695
Epoch [2/5], Step [200/600], Loss: 0.1396
Epoch [2/5], Step [300/600], Loss: 0.0869
Epoch [2/5], Step [400/600], Loss: 0.1663
Epoch [2/5], Step [500/600], Loss: 0.1305
Epoch [2/5], Step [600/600], Loss: 0.0850
Epoch [3/5], Step [100/600], Loss: 0.0793
Epoch [3/5], Step [200/600], Loss: 0.0579
Epoch [3/5], Step [300/600], Loss: 0.0778
Epoch [3/5], Step [400/600], Loss: 0.0564
Epoch [3/5], Step [500/600], Loss: 0.1795
Epoch [3/5], Step [600/600], Loss: 0.0422
Epoch [4/5], Step [100/600], Loss: 0.0360
Epoch [4/5], Step [200/600], Loss: 0.1245
Epoch [4/5], Step [300/600], Loss: 0.1650
Epoch [4/5], Step [400/600], Loss: 0.0849
Epoch [4/5], Step [500/600], Loss: 0.0566
Epoch [4/5], Step [600/600], Loss:

#Q3 (20 points): Discuss the results. Is the neural network doing a good job?

In [12]:
#@title Evaluating the accuracy of the model

correct = 0
total = 0
for images,labels in test_gen:
  images = Variable(images.view(-1,28*28))
  labels = labels

  output = net(images)
  _, predicted = torch.max(output,1)
  correct += (predicted == labels).sum()
  total += labels.size(0)

print('Accuracy of the model: %.3f %%' %((100*correct)/(total+1)))
#Q3 (20 points): How to interpret the results? Is the neural network does a good job?

Accuracy of the model: 97.910 %


The accuracy determines how well the model performs to unseen data. High accuracy indicates that the model does a good job of generalizing from the training data to the unseen data. But if the accuracy is low, then it indicates that the model does not do a good job at doing this. It could be due to issues such as underfitting or overfitting. Similarly, if the loss decreases during the training it indicates that the model is learning. 

In this example, the neural network is doing a good job. The accuracy of the model is high at 97.91%. Similarly as seen throughout the training of the model, there is a tend of a decreasing loss indicating that the model is learning efficiently. 